src/unit.h

```diff
@@ -6,25 +6,162 @@
+struct wavelength_base_unit
+    : boost::units::base_unit<wavelength_base_unit,
+                              boost::units::length_dimension, 2> {};
+struct energy_base_unit
+    : boost::units::base_unit<energy_base_unit, boost::units::energy_dimension,
+                              3> {};
+struct tof_base_unit
+    : boost::units::base_unit<tof_base_unit, boost::units::time_dimension, 4> {
+};
+
+// Create the units system using the make_system utility
+typedef boost::units::make_system<wavelength_base_unit, counts_base_unit,
+                                  energy_base_unit, tof_base_unit>::type
+    units_system;
+
+typedef boost::units::unit<counts_dimension, units_system> counts;
+typedef boost::units::unit<boost::units::length_dimension, units_system>
+    wavelength;
+typedef boost::units::unit<boost::units::energy_dimension, units_system> energy;
+typedef boost::units::unit<boost::units::time_dimension, units_system> tof;
+
+/// unit constants
+BOOST_UNITS_STATIC_CONSTANT(angstrom, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(angstroms, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(meV, energy);
+BOOST_UNITS_STATIC_CONSTANT(meVs, energy);
+BOOST_UNITS_STATIC_CONSTANT(microsecond, tof);
+BOOST_UNITS_STATIC_CONSTANT(microseconds, tof);
+
+} // namespace tof
+} // namespace neutron
+
+// Define conversion factors: the conversion will work both ways with a single
+// macro
+
+// Convert angstroms to SI meters
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(neutron::tof::wavelength_base_unit,
+                                     boost::units::si::length, double, 1.0e-10);
+// Convert meV to SI Joule
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(
+    neutron::tof::energy_base_unit, boost::units::si::energy, double,
+    1.0e-3 * boost::units::si::constants::codata::e.value().value());
+// Convert tof microseconds to SI seconds
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(neutron::tof::tof_base_unit,
+                                     boost::units::si::time, double, 1.0e-6);
+
+template <>
+struct boost::units::base_unit_info<neutron::tof::wavelength_base_unit> {
+  static std::string name() { return "angstroms"; }
+  static std::string symbol() { return "AA"; }
+};
+
+template <>
+struct boost::units::base_unit_info<neutron::tof::counts_base_unit> {
+  static std::string name() { return "counts"; }
+};
+
+template <>
+struct boost::units::base_unit_info<neutron::tof::energy_base_unit> {
+  static std::string name() { return "milli-electronvolt"; }
+  static std::string symbol() { return "meV"; }
+};
+
+template <> struct boost::units::base_unit_info<neutron::tof::tof_base_unit> {
+  static std::string name() { return "microseconds"; }
+  static std::string symbol() { return "us"; }
+};
+
+namespace {
+boost::units::si::dimensionless none;
+boost::units::si::length m;
+boost::units::si::area m2;
+boost::units::si::time s;
+boost::units::si::mass kg;
+neutron::tof::counts counts;
+neutron::tof::wavelength lambda;
+neutron::tof::energy mev;
+neutron::tof::tof tof;
```

Comment:
I'd make these `constexpr`.

Reply:
with `constexpr neutron::tof::tof tof;`, i'm getting `error: the type of constexpr variable ‘{anonymous}::tof’ is not literal`

Reply:
But it works with the others?

Reply:
no, does not work for any of them. That was just an example.

Reply:
Ok, this seems to indicate that the official `boost::units` does not support `constexpr` (but it was apparently merge into a developement branch?): https://github.com/boostorg/units/pull/22